modules/stochastic_tools/src/reporters/ActiveLearningGPDecision.C

```diff
@@ -0,0 +1,239 @@
+  else if (function_name == "COV")
+  {
+    f_val = gp_std / std::abs(gp_mean);
+    if (f_val < threshold)
+      result = 1;
+  }
+  else
+    ::mooseError("Invalid learning function ", std::string(function_name));
+  return result;
+}
+
+void
+ActiveLearningGPDecision::setupData(const std::vector<Real> & output_parallel,
+                                    const std::vector<std::vector<Real>> & inputs_prev)
+{
+  for (dof_id_type ss = 0; ss < _sampler.getNumberOfRows(); ++ss)
+  {
+    _outputs_batch.push_back(output_parallel[ss]);
+    for (unsigned int k = 0; k < _sampler.getNumberOfCols(); ++k)
+      _inputs_batch[k].push_back(inputs_prev[ss][k]);
+  }
+}
+
+void
+ActiveLearningGPDecision::facilitateDecision(const std::vector<Real> & row,
+                                             dof_id_type local_ind,
+                                             Real & val,
+                                             const bool & retrain)
+{
+  _gp_sto[0] = _gp_eval->evaluate(row, _gp_sto[1]);
+  _gp_sto[1] = retrain ? 0.0 : _gp_sto[1];
+  const bool lf_indicator = learningFunction(_gp_sto[0],
+                                             _gp_sto[1],
+                                             _learning_function,
+                                             *_learning_function_parameter,
+                                             _learning_function_threshold);
+  if (_flag_sample[local_ind] == true)
+    _flag_sample[local_ind] = false;
+  if (lf_indicator)
+  {
+    val = _gp_sto[0];
+    _decision[local_ind] = false;
+  }
+  else
+  {
+    ++_track_gp_fails;
+    _flag_sample[local_ind] = true;
+  }
+  _gp_mean_parallel[local_ind] = _gp_sto[0];
+  _gp_std_parallel[local_ind] = _flag_sample[local_ind] == true ? 0.0 : _gp_sto[1];
+}
+
+void
+ActiveLearningGPDecision::transferOutput(const DenseMatrix<Real> & inputs_parallel,
+                                         const std::vector<Real> & gp_mean_parallel,
+                                         const std::vector<Real> & gp_std_parallel)
+{
+  for (dof_id_type ss = 0; ss < _sampler.getNumberOfRows(); ++ss)
+  {
+    for (unsigned int j = 0; j < _sampler.getNumberOfCols(); ++j)
+      _inputs[ss][j] = inputs_parallel(ss, j);
+    _gp_mean[ss] = gp_mean_parallel[ss];
+    _gp_std[ss] = gp_std_parallel[ss];
+  }
+}
+
+bool
+ActiveLearningGPDecision::needSample(const std::vector<Real> & row,
+                                     dof_id_type local_ind,
+                                     dof_id_type,
+                                     Real & val)
+{
+  _gp_sto.resize(2);
+  _output_parallel.resize(1);
+  _gp_mean_parallel.resize(1);
+  _gp_std_parallel.resize(1);
+  DenseMatrix<Real> inputs_parallel(_sampler.getNumberOfRows(), _sampler.getNumberOfCols());
+  // Storing all the inputs and outputs (required for each time step of the mainApp)
+  for (dof_id_type ss = _sampler.getLocalRowBegin(); ss < _sampler.getLocalRowEnd(); ++ss)
+  {
+    const auto data = _sampler.getNextLocalRow();
```

Comment:
I am very concerned by this call. The base class is also calling it in the `needSample` loop, which will cause some unexpected behavior. My suggestion is to do all this within the `initialize` function. Then use this `needSample` to just communicate the results to the base class.

Reply:
I have a question about this. We are overriding the needSample function right? So, why would the base class still call it? Anyhow, one quick fix is to throw an error whenever the number of procs are greater than the number of rows. This would be an unnecessary expenditure of resources, right?